components/dm/spec/grafana.go

```diff
@@ -0,0 +1,227 @@
+// Copyright 2020 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package spec
+
+import (
+	"fmt"
+	"path/filepath"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tiup/pkg/cluster"
+	"github.com/pingcap/tiup/pkg/cluster/executor"
+	"github.com/pingcap/tiup/pkg/cluster/spec"
+	"github.com/pingcap/tiup/pkg/cluster/task"
+	"github.com/pingcap/tiup/pkg/cluster/template/config"
+	"github.com/pingcap/tiup/pkg/cluster/template/scripts"
+	"github.com/pingcap/tiup/pkg/meta"
+)
+
+// GrafanaComponent represents Grafana component.
+type GrafanaComponent struct{ *Topology }
+
+// Name implements Component interface.
+func (c *GrafanaComponent) Name() string {
+	return ComponentGrafana
+}
+
+// Instances implements Component interface.
+func (c *GrafanaComponent) Instances() []Instance {
+	ins := make([]Instance, 0, len(c.Grafana))
+	for _, s := range c.Grafana {
+		ins = append(ins, &GrafanaInstance{
+			BaseInstance: spec.BaseInstance{
+				InstanceSpec: s,
+				Name:         c.Name(),
+				Host:         s.Host,
+				Port:         s.Port,
+				SSHP:         s.SSHPort,
+
+				Ports: []int{
+					s.Port,
+				},
+				Dirs: []string{
+					s.DeployDir,
+				},
+				StatusFn: func(_ ...string) string {
+					return "-"
+				},
+			},
+			topo: c.Topology,
+		})
+	}
+	return ins
+}
+
+// GrafanaInstance represent the grafana instance
+type GrafanaInstance struct {
+	spec.BaseInstance
+	topo *Topology
+}
+
+// InitConfig implement Instance interface
+func (i *GrafanaInstance) InitConfig(e executor.Executor, clusterName, clusterVersion, deployUser string, paths meta.DirPaths) error {
```

Comment:
The logic is the same with https://github.com/pingcap/tiup/blob/master/pkg/cluster/spec/grafana.go#L101

Can they share the code?

Reply:
They different, this init all the config need instead of making directories and copy files from another directory in the run_** script for the cluster one, which I think it's obscure.